app/routes/admin.users.$userId.tsx

```diff
@@ -0,0 +1,124 @@
+/*!
+ * Copyright © 2023 United States Government as represented by the
+ * Administrator of the National Aeronautics and Space Administration.
+ * All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import type { ActionFunctionArgs, LoaderFunctionArgs } from '@remix-run/node'
+import { useFetcher, useLoaderData } from '@remix-run/react'
+import { Button, Checkbox } from '@trussworks/react-uswds'
+
+import { getUser } from './_auth/user.server'
+import { adminGroup } from './admin'
+import {
+  addUserToGroup,
+  getCognitoUserFromSub,
+  getGroups,
+  listGroupsForUser,
+  removeUserFromGroup,
+} from '~/lib/cognito.server'
+
+interface GroupSelectionItem {
+  groupName: string
+  defaultChecked: boolean
+  description: string
+}
+
+export async function loader({
+  params: { userId },
+  request,
+}: LoaderFunctionArgs) {
+  const currentUser = await getUser(request)
+  if (!currentUser?.groups.includes(adminGroup))
+    throw new Response(null, { status: 403 })
+  if (!userId) throw new Response(null, { status: 404 })
+  const user = await getCognitoUserFromSub(userId)
+  const userGroups = (await listGroupsForUser(userId)).map(
+    (group) => group.GroupName
+  )
+  const allGroups: GroupSelectionItem[] = (await getGroups())
+    .map((x) => {
+      return {
+        groupName: x.GroupName ?? '',
+        defaultChecked: userGroups.includes(x.GroupName),
+        description: x.Description ?? '',
+      }
+    })
+    .filter((x) => Boolean(x.groupName))
+  return { user, allGroups }
+}
+
+export async function action({
+  request,
+  params: { userId },
+}: ActionFunctionArgs) {
+  const user = await getUser(request)
+  if (!user?.groups.includes(adminGroup))
+    throw new Response(null, { status: 403 })
+  const data = await request.formData()
+  const { ...selectedGroups } = Object.fromEntries(data)
+  if (!userId) throw new Response(null, { status: 400 })
+  const currentUserGroups = (await listGroupsForUser(userId)).map(
+    (x) => x.GroupName
+  ) as string[]
+  const selectedGroupsNames = Object.keys(selectedGroups)
+
+  await Promise.all([
+    ...selectedGroupsNames
+      .filter((x) => !currentUserGroups.includes(x))
+      .map((x) => addUserToGroup(userId, x)),
+    ...currentUserGroups
+      .filter((x) => !selectedGroupsNames.includes(x))
+      .map((x) => removeUserFromGroup(userId, x)),
```

Comment:
There is some missing validation here. Submitting this form must not add or remove the special groups that Cognito uses to keep track of which users federated how.